### DISCUSSION TITLE
Relationship between `extract_constants`, `read_constants` and `interpolate_constants`

### DISCUSSION THREAD

**Discussion by robbibt** [pyTMD/pyTMD#355] [General]
Hi @tsutterley, I'm finally looking into refactoring some of our pyTMD tide modelling code to try and minimise IO by loading consistuents one time at the beginning of an analysis, interpolating values from those loaded consistuents in parallel, then model tides from those interpolated constituents as a final step.

It looks like `model.read_constants` and `model.interpolate_constants` are what I want, but I just wanted to check a few assumptions first:

1) Are `read_constants` and `interpolate_constants` in combination identical to `extract_constants`? e.g. are the following two equivelent? (in the code I can see that `extract_constants` [doesn't call `read_constants` and `interpolate_constants`](https://github.com/tsutterley/pyTMD/blob/main/pyTMD/io/model.py#L967-L1058) internally):
```
#1
amplitude, phase = extract_constants()

#2
c = read_constants()
amplitude, phase = interpolate_constants(c)
```
2) Do `read_constants` and `interpolate_constants` support the same overall params/functionality as `extract_constants`? I'm particularly thinking for things like `crop`, `bounds` etc - ideally I'd love to use `read_constants` to read in cropped constituents for my study area, then use that cropped data for the interpolation via `interpolate_constants` to keep run times quick.

**Comment by robbibt** [comment]
Actually, while looking into this I may have found a bug in the new `model.extract_constants` method; raised here: #356

**Comment by tsutterley** [comment]
Hey @robbibt

1) they should give identical answers and there are tests (e.g. [here](https://github.com/tsutterley/pyTMD/blob/main/test/test_download_and_read.py#L604-L676)) to verify that they are consistent. The major difference between the workflows is that using `read_constants` and `interpolate_constants` will keep the (possibly cropped) constituents in memory. This can reduce computational time in the case where the same grids will be reused (no need to reread the data). `extract_constants` does not cache any of the constituent grids and so can be faster in the case where the data is only used once.
2) They should support the same overall functionality (e.g. `crop` and `bounds`). The read and reapply workflow is the target use case for `read_constants` and `interpolate_constants`.

**Reply by robbibt** [reply]
Fantastic, thanks @tsutterley - I have been meaning to refactor our workflows to use these sub-functions for a while but just wanted to check if my assumptions were all correct before starting on it. This sounds like it should do exactly what I am hoping for!
